### ISSUE TITLE
Move proxy registrations in reflection config starting with GraalVM / Mandrel for JDK 23 (24.1)

### ISSUE THREAD

**Issue by zakkak** [quarkusio/quarkus#41016]
### Description

Starting with GraalVM / Mandrel for JDK 23 (24.1) the `proxy-config.json` file will be deprecated. Instead, proxy registrations should be moved to the reflection config, see https://github.com/oracle/graal/pull/8822

**Comment by jerboaa** [comment]
> Ideally we should only generate proxy-config.json when using GraalVM / Mandrel < 24.1 and reachability-metadata.json when using >= 24.1

How do the files differ? If they don't differ we could generate `proxy-config.json` and in the native build step rename to `reachability-metadata.json` (if the version is >= 24.1) or some such before native image runs? Just a thought.

**Comment by zakkak** [comment]
The new file follows a [new schema](https://github.com/oracle/graal/pull/9048/files#diff-0c6dfb22780a72a74c06a029c09a0ac287ce35a428120088ed60873081c61f38) that essentially combines all previous *-config.json files in one file so a simple rename wouldn't suffice.

An adapter that would merge the multiple config files in one could be an option but feels hacky to me.

**Comment by zakkak** [comment]
It looks like there are considerations to backport `reachability-metadata.json` parsing to GraalVM 23.1 for JDK 21 (https://github.com/oracle/graal/pull/9606) which could make the transition a bit easier.

**Comment by zakkak** [comment]
FTR backwards compatibility has been merged in Mandrel/GraalVM for JDK 21 (see https://github.com/graalvm/graalvm-community-jdk21u/pull/23) however, the schema doens't seem very stable yet, so I suggest we delay the transition till it becomes more stable (probably after the GraalVM for JDK 25 release).

**Comment by zakkak** [comment]
@gsmet I am planning to implement this once we move to Mandrel for JDK 25 (in late September / early October?). 

This "new" json format is already supported in Mandrel for JDK 21 so in theory we could backport the changes to earlier Quarkus LTS releases, but this might be a big change for users. WDYT?

**Comment by djarnis73** [comment]
Currently the [Native Image Tracing Agent Integration](https://quarkus.io/guides/native-reference#native-image-agent-integration) feature is broken due to this if you try it with mandrel 24.1 or later.

Because the agent generates a `reachability-metadata.json` and the agent integration expects to find the older set of json files.

It would be nice if it could read the `reachability-metadata.json` and still generate the old style files since the newer graalvm version can still read the old style files.

**Comment by zakkak** [comment]
@djarnis73 thanks for raising this. It sounds like a separate issue though so I forked it in https://github.com/quarkusio/quarkus/issues/50294

**Comment by zakkak** [comment]
This is blocked by https://github.com/graalvm/graalvm-community-jdk21u/issues/217

To ensure backwards compatibility we need `GraalVM/Mandrel 23.1 for JDK 21` to support lambda reflection.

**Comment by rsvoboda** [comment]
@zakkak I noticed `Warning: Option 'DynamicProxyConfigurationResources' is deprecated and might be removed in a future release: This can be caused by a proxy-config.json file in your META-INF directory. Consider including proxy configuration in the reflection section of reachability-metadata.md instead.. Please refer to the GraalVM release notes.` warning with 3.31.1 which uses Java 25 based Mandrel image.

What's the current situation with this issue, last comment is from October 2025?

**Comment by zakkak** [comment]
Hi @rsvoboda GraalVM/Mandrel 23.1 for JDK 21 [still doesn't include the necessary support for `reachability-metadata.md`](https://github.com/graalvm/graalvm-community-jdk21u/issues/217), this will probably come in the next CPU release (in April 2026).

Our plan is to start emitting both the new `reachability-metadata.md` file along with the old `*-config.json` files for things that GraalVM/Mandrel 23.1 doesn't support. Note however that this might still result in some different warning.
